Rewrite asserts in test-modules loaded very early in the startup

nicoddemus opened the pull request:
I investigated this some more and found the problem:

The assertion-rewrite hook is installed very early on, during the construction of the `Config` object. After installing the hook and still during `Config`'s construction, plugins and `conftest` files are also imported and are rewritten if `AssertionRewriteHook._should_rewrite` returns `True`. 

Now the problem with the example posted in #1784 is that it imports `test_foocompare` at the top-level of the `conftest.py` file which is loaded very early, when the `session` object has not even been created yet (much less set on the hook), so it doesn't try to rewrite that module.

Looking at the code I noticed that the code block that should handle this, the `for pat in self.fnpaths` loop, doesn't really depend on `self.session` at all except for that weird comment about a cycle. Changing the code to always check if the filename matches one of the test patterns seems to do the trick.

Fix #1784

nicoddemus commented:
cc @flub

flub commented on _pytest/assertion/rewrite.py:
```diff
@@ -144,28 +145,24 @@ def _should_rewrite(self, name, fn_pypath, state):
+        # modules not passed explicitly on the command line are only
+        # rewritten if they match the naming convention for test files
+        for pat in self.fnpats:
+            if fnmatch(fn_pypath.basename, pat):
```
Why the switch away from `fn_pypath.fnmatch(pat)`?  Does this have to do something with the session thing?

nicoddemus replied:
`fn_pypath.fnmatch(pat)` would trigger an `fnmatch` import, which would in turn cause this method to be called recursively until it blew up the stack.

flub replied:
Ah, so that's the recursive bit then!  I'd be in favour of putting that exact sentence in the comment just before, I think it's non obvious and warrants explaining.

nicoddemus replied:
Done, thanks!

flub commented on _pytest/assertion/rewrite.py:
```diff
@@ -144,28 +145,24 @@ def _should_rewrite(self, name, fn_pypath, state):
+
+        for marked in self._must_rewrite:
+            if marked.startswith(name):
+                return True
```
Not really this PR, but I guess we should be doing an `state.trace("matched marked file %r" % name)` here to be consistent.

nicoddemus replied:
Hmm just noticed this, but shouldn't this be `if name.startswith(marked):`? I'm supposing `self._must_rewrite` is a list of module prefixes like `"pytest_qt"`, intended to match `pytest_qt` and any submodule.

nicoddemus replied:
Anyways, added the trace call, thanks. :grin:

flub commented:
Looks good to me, nice debugging!

nicoddemus commented on _pytest/assertion/rewrite.py:
```diff
@@ -144,28 +145,25 @@ def _should_rewrite(self, name, fn_pypath, state):
+                return True
+
+        for marked in self._must_rewrite:
+            if marked.startswith(name):
```
Hmm just noticed this, but shouldn't this be `if name.startswith(marked):`? I'm supposing `self._must_rewrite` is a list of module prefixes like `"pytest_qt"`, intended to match `pytest_qt` and any submodule.

flub replied:
The `_must_rewrite` set can also be full module paths, like `myplugin.plugin`.  This way round `myplugin.plugin.foo` or `myplugin.plugin.__init__` would not be rewritten but if it was the other way round it would be.  I guess it could just be a simple `==` comparison though?

flub replied:
However thinking about it some more maybe `name.startswith(marked)` would be better behaviour.  We should probably also update the docstring in `assertion/__init__.py;register_assert_rewrite` to read "will make sure that this module or all modules inside the package will get their assert statements rewritten" in that case though.

The only downside of that way round that I can think of is that someone may want to register `mypkg.__init__` explicitly rather then just `mypkg`, but that's probably ok.

nicoddemus commented:
Implemented all of @flub's suggestions (thanks!) and squashed the commits. I will merge this later unless anybody else wants more time to review this further. :grin:
